### DISCUSSION TITLE
keybindings should take precedence over standard menu shortcuts

### DISCUSSION THREAD

**Discussion by cortesi** [ghostty-org/ghostty#3187] [Feature Requests, Ideas] [answered]
For example, bind a key like this:

```
keybind = cmd+h=goto_split:left
```

Observe that `cmd-h` still hides the window, which is the default menu shortcut, rather than taking the configured action.

**Comment by mitchellh** [comment]
These are system level keybinds defined in `System Settings.app`. Is there a way to override these? I'm open to it if its possible but from what I've found so far its pretty hard to do unless you fully remove the menu item.

**Reply by cortesi** [reply]
I'm not a Swift developer, but it looks to me like these bindings are defined in the user interface using the "keyEquivalent" attribute:

https://github.com/ghostty-org/ghostty/blob/35b9ceee2116331b83c0c86269394e2545070b0f/macos/Sources/App/macOS/MainMenu.xib#L94

So, I believe the bindings are just a convention not a system global setting, and that Ghostty hard-binds a lot of shortcut keys to menu items that users would plausibly want to over-ride. If individually over-riding key bindings would be fiddly, I'd be happy with a global option that turns off all menu shortcuts.

**Reply by mitchellh** [reply]
Those are defaults but we override them all here: https://github.com/ghostty-org/ghostty/blob/35b9ceee2116331b83c0c86269394e2545070b0f/macos/Sources/App/macOS/AppDelegate.swift#L339

Let me see if we sync that one.

**Reply by lalitkapoor** [reply]
I'm looking to repurpose `cmd+[` and `cmd+]` and I believe it is also impacted by this. 

Strangely, even using the shift key modifier like so doesn't seem to work `keybind = cmd+shift+[=text:h`

**Reply by radiaku** [reply]
Same here, I looking to repurpose `cmd+h` to `C+w+h` and `cmd+l `to `C+w+l` . This my keybinding accros mac, linux and windows for my vim purpose.

**Reply by ThiefZero** [reply]
Tried all different kinds of ways to rebind `ctrl+tab` and failed. Same issue as the other keybinds mentioned I believe, as the default behaviour in ghostty is to switch to the next tab.

**Reply by alinalihassan** [reply]
This still keeps me from fully adopting Ghostty, I got used to using cmd+hjkl to move around panes in WezTerm, but I don't know how to override them or where to look in the repo for that.

**Reply by aaronjensen** [reply]
For me, it's `cmd+m`, ah and ctrl+tab/ctrl+shift+tab

**Reply by lalitkapoor** [reply]
I was able to bind to `cmd+shift+[` and `cmd+shift+]`. I saw from the terminal inspector that this is actually `left_bracket` and `right_bracket` instead of `[` and `]`.

```
keybind = cmd+shift+left_bracket=csi:2;10[
keybind = cmd+shift+right_bracket=csi:2;10]
```

**Comment by IanLiuTW** [comment]
**A quick workaround** 
If you want to avoid `cmd+h` from hiding the window on MacOS, you can go to `Keyboard Shortcuts` in the system settings and override it with other keys for Ghostty. This would free the default `cmd+h` binding. It works well for me, and it doesn't require installing and configuring additional apps.

![image](https://github.com/user-attachments/assets/6ae573a6-8a7f-4ade-b956-6d7089b91cf8)

For me, I am having issue with `super+.` \ `super+period` on MacOS. Seems like this combination is ignored or blocked on system level. I confirmed it using the inspector. 

P.S. The `super+shift+[` and `super+shift+]` on MacOS are also used to go to previous/next tabs b default. I think it's on the system level since you can observe the same behavious in browsers and Finder, and there's no way, as far as I know, to rebind them in system settings. This makes those bindings unavailable.

**Reply by lalitkapoor** [reply]
I was able to solve for the square brackets - https://github.com/ghostty-org/ghostty/discussions/3187#discussioncomment-11681896

and in my tmux config I was able to bind to it like so:

```
set -s user-keys[2] "\e[2;10]"  # Cmd+Shift+[
set -s user-keys[3] "\e[2;10["  # Cmd+Shift+]

bind -n User2 select-window -t -1
bind -n User3 select-window -t +1
```

**Reply by IanLiuTW** [reply]
Sorry for the confusion. In my previous config, I also tried
```
keybind = cmd+shift+left_bracket=move_tab:-1
keybind = cmd+shift+right_bracket=move_tab:+1
```
and ran into the issue of triggering the system level bindings first.

**Reply by arnaudlamy** [reply]
thx!!

**Comment by mitchellh** [comment]
Okay, this has a lot of upvotes. I'm completely willing to accept it, but I want to once again just want to do some research into how...

**Reply by aaronjensen** [reply]
FWIW, iTerm2 supports this. They also show cmd+M in the menu item still, but their keybindings take precedence over the menu keybindings. I mention this in case it helps your research effort...

**Reply by milch** [reply]
I just looked into this and it seems that there are 2 issues here with different root causes. There's the menu shortcut preference issue which people run into when they try to bind "Cmd+h", and there's another issue which is what people run into when they try to bind `]` or something like `Cmd+1`:

1. When you set up a keybinding using a key where the symbol is different than the name defined in `key.zig` [here](https://github.com/ghostty-org/ghostty/blob/c8950d376ab3be13fbd9a19317163075f4feddbc/src/input/key.zig#L287) (e.g. `[` vs. `left_bracket`) this get inserted into the`bindings` map as a unicode code point here: https://github.com/ghostty-org/ghostty/blob/c8950d376ab3be13fbd9a19317163075f4feddbc/src/input/Binding.zig#L1022
2. The defaults are set up using the translated name from the `Key` enum here: https://github.com/ghostty-org/ghostty/blob/c8950d376ab3be13fbd9a19317163075f4feddbc/src/config/Config.zig#L2469
3. When the event makes it back through the native event handler down to the code where it gets looked up in `Bindings.zig`, that lookup happens with the translated name as the highest priority and the unicode code point as the lowest priority: https://github.com/ghostty-org/ghostty/blob/c8950d376ab3be13fbd9a19317163075f4feddbc/src/input/Binding.zig#L1427

i.e. the binding gets saved correctly, makes it through all the intermediate event handlers correctly, but at the final lookup there is a "higher priority" default binding that gets used instead. This is why you can for example bind `cmd+one` but not `cmd+1`

**Reply by milch** [reply]
By the way, for the first issue of rebinding the menu shortcuts, this can be handled by the app delegate. If the event is returned in either of the two branches here then it'll do the default, otherwise it'll suppress it:

https://github.com/ghostty-org/ghostty/blob/c8950d376ab3be13fbd9a19317163075f4feddbc/macos/Sources/App/macOS/AppDelegate.swift#L426-L432

**Reply by rhodes-b** [reply]
~~the literal `1` is a completely separate thing which is not an issue you need to use the spelling of the number so `one` in this case.~~ 

~~The docs werent very clear on this originally but are now a bit more explicit~~
https://ghostty.org/docs/config/reference#keybind 

edit: this does seem to work, it was some of our original understanding it doesn't. I see now that you need to clear the old keybind as the literal 1 and one are treated separately

**Comment by radiaku** [comment]
Okay, I can make this work on neovim. I leave this for those who want to repurpose their keybinding to:

```
cmd+h to C+w+h
cmd+l to C+w+l
```

First change  Keyboard setting ( on OSX sonoma )  , `settings -> keyboard -> keyboard shortcuts -> app shortcuts` and add `Hide  Ghostty
`
<img width="705" alt="Screenshot 2024-12-28 at 11 51 23" src="https://github.com/user-attachments/assets/d0d8abb3-b35e-4224-b390-9b6c38a0d2b9" />


and then the config is

```
macos-option-as-alt = true
keybind = cmd+l=text:\x17\x6C
keybind = cmd+h=text:\x17\x68

```

https://github.com/user-attachments/assets/e9e10445-84a3-4f86-b66c-f9eeb6e2fd4a

Now I can move from iterms to ghostty.

**Reply by LeVuMinhHuy** [reply]
you are my hero, wish you all the best 🙏

**Reply by jibundit** [reply]
Thank you for this suggestion. However it only works the nvim session, and not the shell level, I made some tweaks to work this out.

```
macos-option-as-alt = true
keybind = cmd+h=goto_split:left
keybind = cmd+j=goto_split:bottom
keybind = cmd+k=goto_split:top
keybind = cmd+l=goto_split:right
```

**Comment by frobware** [comment]
In https://github.com/ghostty-org/ghostty/discussions/4134 I was using /usr/bin/defaults to clear out menu bar shortcuts. I don't know off-hand if you can do that programatically.

**Comment by mitchellh** [comment, marked as answer]
I figured out how to do this. Issue #4590

**Reply by renxzen** [reply]
Will this NSEvent solution also work for ctrl+tab or ctrl+shift+tab?

**Reply by mitchellh** [reply]
It should but I will verify.

**Reply by ThiefZero** [reply]
can confirm that adding ```keybind = ctrl+tab=text:\x1b[27;5;9~``` to `config` works for me.
(still figuring out how to add the `shift+ctrl+tab`)

**Reply by renxzen** [reply]
Thank you @ThiefZero 
Could you please try with:

```
keybind = ctrl+tab=esc:[27;5;9~
keybind = ctrl+shift+tab=esc:[27;6;9~
```

Which is what works for me in linux

**Reply by ThiefZero** [reply]
@renxzen 
Thanks for your reply.
`keybind = ctrl+tab=esc:[27;5;9~` works!
`keybind = ctrl+shift+tab=esc:[27;6;9~` can't seem to get this one to work. Trying to capture it with `tmux`, but instead `nvim` sees it as `shift+tab`. Could be my config, though the same setup I had works in `kitty`.

**Reply by renxzen** [reply]
finally could use my work mac to test the changes.
with the same configs i stated above, i could get it to work for tmux without issues

```tmux
bind -n C-Tab next-window
bind -n C-BTab previous-window
```

previously unbinding them on tmux, i also tested for your nvim case @ThiefZero and it also worked without issues

```lua
vim.api.nvim_set_keymap('n', '<C-S-Tab>', '<cmd>tabp<cr>', { noremap = true, silent = true })
vim.api.nvim_set_keymap('n', '<C-Tab>', '<cmd>tabn<cr>', { noremap = true, silent = true })
```

furthermore, i leave my config here to see if it helps you [gist](https://gist.github.com/renxzen/d9888fc945f4185eb63d91c97d3cfad3)

**Reply by ThiefZero** [reply]
@renxzen 
`bind -n C-BTab previous-window` this did it, thank you! It used to be `bind -n S-C-Tab previous-window` which worked flawlessly in `kitty`, weirdly. But thank you so much for taking your time to help fix this! 😄

**Reply by aaronjensen** [reply]
If you want to use `C-S-Tab`, you can use this:

```
keybind = ctrl+tab=text:\x1b[9;5u
keybind = ctrl+shift+tab=text:\x1b[1;5Z
```

Note that for some reason `esc:[1;5Z` does not work through tmux. It ends up coming through as just backtab I think.

Also, afaict, this is the correct escape sequence for ctrl-tab and ctrl-shift-tab per fixterms: https://www.leonerd.org.uk/hacks/fixterms/

I don't know what the other one represents. Please correct me if I'm wrong.

**Comment by zoriya** [comment]
Could `keybind = clear` also clear those OS level shortcuts? or do we need to manually add a `keybind = cmd+h=unbind`?
